Add nethealth container to prepull manifest

girishkalele opened the pull request.

Commit:
Add nethealth prepull container output to e2e run logs

girishkalele commented:
https://github.com/kubernetes/kubernetes/issues/26626

dchen1107 commented:
@girishkalele I think I didn't make the suggestion clear at #26626. What I really suggested is introducing another container paralleling with image-puller container to run your nethealth in prepull pod since that Pod is one of the first pod running at test nodes. 

Your current implementation only pulling the image, but not run the test.

bprashanth commented:
I'm not sure why we'd want to prepull if we're going to add this as a container in the prepull pod anyway. It's just dead weight no?

girishkalele commented:
@dchen1107 

I misunderstood your comment about adding to the manifest - still new here - I have now added a new container named nethealth-check and removed it from the pre-pull image list.
Will the output of this container also show up in the node logs ?
I need to specify the container name with a -c to view its logs (since the logs command defaults to the image puller container).

```console
$ kubectl logs --namespace=kube-system e2e-image-puller2 -c nethealth-check
2016/06/10 17:48:18 HTTP HEAD reports content length: 67108864 - running GET
2016/06/10 17:48:19 DOWNLOAD: 67108864 bytes 1103 ms Bandwidth ~ 59416 KiB/sec
2016/06/10 17:48:19 Hash Matches expected value
```

bprashanth commented:
This is what prints it:https://github.com/kubernetes/kubernetes/blob/cf234ab67d3aaf58c4dcf3eef38758e1ae986d67/test/e2e/framework/framework.go#L245, it should iterate through all containers (https://github.com/kubernetes/kubernetes/blob/66149c50fd00550e55920d5e78f7dbc5e4283127/test/e2e/framework/util.go#L660). You can trigger a test failure and confirm that it prints the output of your container.

Where's the code for kube-nethealth? It would be nice if it gathered background stats while the image puller is running vs just fire and forget a single curl request.

bprashanth commented on cluster/saltbase/salt/e2e-image-puller/e2e-image-puller.manifest:
```diff
@@ -35,6 +35,14 @@ spec:
+        cpu: 100m
+      limits:
+        cpu: 100m
+    image:  gcr.io/google_containers/kube-nethealth-amd64:1.0
```
this needs to exit or the pod will continue wasting resources on the node and tests like the resource tracking test will fail waiting for it to stop. Ideally it should exit when the image puller exits. This is why I'd suggested breaking out the for loop in sh -c above into an independent script, and doing your stats collection in between each image pull. 

I'm ok with quick and dirty for now, if your container is just going to do some bounded, small number of queries within eg: 4m, which is roughly how long the image puller will take.

girishkalele replied:
Yes, this is just the seed for a more full featured nethealth monitoring container.

girishkalele commented:
Nethealth Code is in the node-problem-detector repo here.

https://github.com/kubernetes/node-problem-detector/blob/master/pkg/util/nethealth/nethealth.go

bprashanth commented:
Defer to Dawn for lgtm, as long as that container does exit in roughly the time the prepuller takes, and logs bw measurements through the functions I pointed to, this is good enough for a first pass

bprashanth commented:
Also worth confirming that the extra cpu it uses doesn't max out the nodes and cause a cluster startup delay, till this container exits and makes room for another cluster addon pod

girishkalele commented:
The container should run with negligible CPU, its doing a GET on a 64MB blob. See time output shows very little user and system CPU runtimes.
Normally exits in ~ 1.5 seconds, depending on bandwidth, but it has a 30 second watchdog timeout that will kill if downloads are abnormally slow.

```console
$ time ./nethealth 
real	0m1.580s
user	0m0.395s
sys	0m0.228s
```

bprashanth commented on cluster/saltbase/salt/e2e-image-puller/e2e-image-puller.manifest:
```diff
@@ -35,6 +35,14 @@ spec:
+  - name: nethealth-check
+    resources:
+      requests:
+        cpu: 100m
```
I guess you're setting 100m 100m to get guaranteed qos, this is what ia was asking for confirmation about (that this 100m won't eject any other pods)

girishkalele replied:
Ah, got it, I used the same resource limits as the image-puller container. This is a short lived container, max 30 seconds; it should vacate its reserved CPU before other pods are scheduled.

bprashanth commented:
If this container crashes its going to leave the pod in an Error state, which will affect the resource usage tracking test (and probably others that simply poll on ready/running/success). Can you confirm that it won't? And if it will crash (non zero exit code), can you make it exit gracefully instead?

girishkalele commented:
@bprashanth 

The nethealth utility was actually made to be scriptable and returns a non-zero return code upon failure. This will leave the container/pod in Error state on nodes with extremely slow networking flagging the error condition in the e2e test logs.
The resource usage tracking though I am not familiar with, I will need to ask the scheduler folks what happens to the resources of a container in Errored state.

bprashanth commented:
e2es will break in various ways if you leave a system pod in an error state. Many of them poll for all system pods in success/ready/running in BeforeEah.

bprashanth commented:
This is why the image prepuller has an exit 0 at the end of the for loop.

girishkalele commented:
@bprashanth 

Ok, I've put in an always exit 0 in the container spec...testing in the deliberate failure case shows the pod successfully finished.

dchen1107 commented:
Can we have the output of nethealth is printed all the time no matter the pod is succeeded or failed?

girishkalele commented:
The nethealth container will always print the b/w measured.
What we need is to change the e2e tests to scan through all image puller:nethealth containers and print the logs even in the success case. I will take a look at the e2e framework.

dchen1107 commented:
I lowered the priority of this pr since the flakiness caused by the network connectivity were reduced dramatically lately. But the stats collected by this test containers / utility are still very important, which can help us to analyze the future failures in a long run.

girishkalele commented:
@k8s-bot Test this again.

fejta commented:
@k8s-bot node e2e test this issue #24479

dchen1107 commented on test/e2e/framework/framework.go:
```diff
@@ -266,6 +266,13 @@ func (f *Framework) AfterEach() {
 		LogPodsWithLabels(f.Client, api.NamespaceSystem, ImagePullerLabels)
```
Since you introduced a new container in this helper pod, is LogPodsWithLabels method still dumping the prepulled images on the node onFailure? Or we should change this method to something like LogContainersInPodsWithLabels(..., "prepuller")?

cc/ @bprashanth

girishkalele replied:
@dchen1107 Good point. Actually, I was thinking we should move the whole print-prepull-container-logs from being printed AfterEach failure to be printed only once, just like the nethealth container.
Not sure if we need the output of the prepull containers after every failed test.

dchen1107 replied:
I agreed with you, @bprashanth WDYT?

bprashanth replied:
I think that was the intent. It was explicity asked for in the original pr, because the prepuller might not finish before we start running tests, and if a test times out image pulling we want to know that we didn't pull that image so we can update the list. I don't have a strong preference eitherway, I don't think we lose anything by being verbose on failure.

dchen1107 commented:
LGTM overall except one nit / question above.

bprashanth commented on test/e2e/framework/framework.go:
```diff
@@ -266,6 +266,13 @@ func (f *Framework) AfterEach() {
+	if TestContext.DumpNethealthLogs {
+		By("Dumping network health container logs from all nodes")
+		LogContainersInPodsWithLabels(f.Client, api.NamespaceSystem, ImagePullerLabels, "nethealth")
+		TestContext.DumpNethealthLogs = false
```
Maybe just do this in https://github.com/kubernetes/kubernetes/blob/master/test/e2e/e2e.go#L145, inconditionally, if you only want it once?

girishkalele replied:
Yes, let me move it there.

bprashanth commented on test/e2e/framework/test_context.go:
```diff
@@ -128,4 +129,5 @@ func RegisterFlags() {
 	flag.BoolVar(&TestContext.GatherMetricsAfterTest, "gather-metrics-at-teardown", false, "If set to true framwork will gather metrics from all components after each test.")
 	flag.StringVar(&TestContext.OutputPrintType, "output-print-type", "hr", "Comma separated list: 'hr' for human readable summaries 'json' for JSON ones.")
 	flag.BoolVar(&TestContext.DumpLogsOnFailure, "dump-logs-on-failure", true, "If set to true test will dump data about the namespace in which test was running.")
+	flag.BoolVar(&TestContext.DumpNethealthLogs, "dump-nethealth-logs", true, "If set to true test will dump data about the network health checks from all nodes.")
```
do we need a new flag for this?

girishkalele replied:
I added the flag (default is true) to follow the same model as DumpLogsOnFailure - looks like there is a use-case for not printing logs sometimes.

bprashanth replied:
nit: can we reuse the same flag dump-logs-on-failure? flag overload is not yet an issue with this binary, but it might be one day like it is in the kubelet/apiserver (100+ flags, old flags are preserved because codepaths have diverged). Node team people will happily tell you what a big problem flag fragmentation is in the internal kubelet equivalent :)

girishkalele replied:
Ok, will do.

bprashanth commented on test/e2e/framework/util.go:
```diff
@@ -693,7 +696,18 @@ func LogPodsWithLabels(c *client.Client, ns string, match map[string]string) {
 	Logf("Running kubectl logs on pods with labels %v in %v", match, ns)
 	for _, pod := range podList.Items {
-		kubectlLogPod(c, pod)
+		kubectlLogPod(c, pod, "")
```
nit: why not just pipe through containerSubstr into this func? I only see 2 uses of it.

girishkalele replied:
I didn't understand your comment - iterate through containers in the pod and call the function with every container ?

bprashanth commented:
lgtm, assuming @dchen1107 is ok with this applying label

girishkalele commented:
@k8s-bot test this again

girishkalele commented:
@k8s-bot test this issue: #27335

bprashanth commented:
Hmm the last failure looks like exactly the sort of things your pr is meant to help with:
```
Jun 15 10:34:16.379: INFO: At {2016-06-15 10:29:45 -0700 PDT} - event for proxy-service-kir7r-fedi4: {kubelet e2e-gce-agent-pr-8-0-minion-group-57ju} Failed: Failed to pull image "gcr.io/google_containers/porter:cd5cb5791ebaa8641955f0e8c2a9bed669b1eaab": Error pulling image (cd5cb5791ebaa8641955f0e8c2a9bed669b1eaab) from gcr.io/google_containers/porter, Get https://storage.googleapis.com/artifacts.google-containers.appspot.com/containers/images/0bac9e8a78faf81cff38ba54baaf63c3818dce51c1ea24d3a516ab8c1ca70a72/ancestry?access_token=ya29.CkECA5qsDgPRsetAhNEghwM6zpkv1fOnJiMaKhk00WMPQMeS1TQT-gpkvCqkZ6P1wrYKRQxFFnPCGSMpBYa1GAssyw: dial tcp 74.125.129.128:443: i/o timeout
```

Did it dump out enough information?

girishkalele commented:
It did dump out information for that node - networking at startup was good.

```
STEP: Logs of kube-system/e2e-image-puller-e2e-gce-agent-pr-8-0-minion-group-57ju:nethealth-check on node e2e-gce-agent-pr-8-0-minion-group-57ju
Jun 15 10:27:48.079: INFO: nethealth : STARTLOG
2016/06/15 17:23:06 HTTP HEAD reports content length: 67108864 - running GET
2016/06/15 17:23:07 DOWNLOAD: 67108864 bytes 761 ms Bandwidth ~ 86118 KiB/sec
2016/06/15 17:23:07 Hash Matches expected value
```

girishkalele commented:
Image puller logs from that node were full of errors, the image gcr.io/google_containers/porter failed to prepull.

```
STEP: Logs of kube-system/e2e-image-puller-e2e-gce-agent-pr-8-0-minion-group-57ju:image-puller on node e2e-gce-agent-pr-8-0-minion-group-57ju
Jun 15 10:34:17.494: INFO: image-puller : STARTLOG
17:22:20 pulling gcr.io/google_containers/busybox
17:22:50 pulling gcr.io/google_containers/busybox:1.24
Error response from daemon: name invalid: Requested repository does not match bearer token resource: google_containers/busybox
17:22:50 pulling gcr.io/google_containers/dnsutils:e2e
17:22:58 pulling gcr.io/google_containers/eptest:0.1
17:22:59 pulling gcr.io/google_containers/fakegitserver:0.1
17:22:59 pulling gcr.io/google_containers/hostexec:1.2
17:23:05 pulling gcr.io/google_containers/iperf:e2e
17:23:13 pulling gcr.io/google_containers/jessie-dnsutils:e2e
17:23:21 pulling gcr.io/google_containers/liveness:e2e
Error response from daemon: name invalid: Requested repository does not match bearer token resource: google_containers/liveness
17:23:22 pulling gcr.io/google_containers/mounttest:0.2
17:23:22 pulling gcr.io/google_containers/mounttest:0.5
17:23:22 pulling gcr.io/google_containers/mounttest:0.6
17:23:28 pulling gcr.io/google_containers/mounttest-user:0.3
17:23:33 pulling gcr.io/google_containers/netexec:1.4
Error response from daemon: name invalid: Requested repository does not match bearer token resource: google_containers/netexec
17:23:33 pulling gcr.io/google_containers/netexec:1.5
17:23:54 pulling gcr.io/google_containers/nettest:1.7
17:23:55 pulling gcr.io/google_containers/nettest:1.8
Error response from daemon: name invalid: Requested repository does not match bearer token resource: google_containers/nettest
17:23:55 pulling gcr.io/google_containers/nginx:1.7.9
17:24:02 pulling gcr.io/google_containers/nginx-slim:0.5
17:24:08 pulling gcr.io/google_containers/n-way-http:1.0
Error response from daemon: name invalid: Requested repository does not match bearer token resource: google_containers/n-way-http
17:24:08 pulling gcr.io/google_containers/pause:2.0
17:24:09 pulling gcr.io/google_containers/pause-amd64:3.0
Error response from daemon: name invalid: Requested repository does not match bearer token resource: google_containers/pause-amd64
17:24:09 pulling gcr.io/google_containers/porter:cd5cb5791ebaa8641955f0e8c2a9bed669b1eaab
Error response from daemon: name invalid: Requested repository does not match bearer token resource: google_containers/porter
17:24:09 pulling gcr.io/google_containers/portforwardtester:1.0
17:24:09 pulling gcr.io/google_containers/redis:e2e
17:24:29 pulling gcr.io/google_containers/resource_consumer:beta2
17:24:41 pulling gcr.io/google_containers/serve_hostname:v1.4
17:24:42 pulling gcr.io/google_containers/servicelb:0.1
17:25:34 pulling gcr.io/google_containers/test-webserver:e2e
17:25:35 pulling gcr.io/google_containers/ubuntu:14.04
17:25:35 pulling gcr.io/google_containers/update-demo:kitten
17:26:10 pulling gcr.io/google_containers/update-demo:nautilus
Error response from daemon: name invalid: Requested repository does not match bearer token resource: google_containers/update-demo
17:26:10 pulling gcr.io/google_containers/volume-ceph:0.1
Error response from daemon: name invalid: Requested repository does not match bearer token resource: google_containers/volume-ceph
17:26:10 pulling gcr.io/google_containers/volume-gluster:0.2
Error response from daemon: name invalid: Requested repository does not match bearer token resource: google_containers/volume-gluster
17:26:10 pulling gcr.io/google_containers/volume-iscsi:0.1
17:26:23 pulling gcr.io/google_containers/volume-nfs:0.6
Error response from daemon: name invalid: Requested repository does not match bearer token resource: google_containers/volume-nfs
17:26:23 pulling gcr.io/google_containers/volume-rbd:0.1
17:26:54 pulling gcr.io/google_samples/gb-redisslave:v1
Error pulling image (0.1) from gcr.io/google_containers/volume-rbd, Server error: Status 0 while fetching image layer (efda3e1d78611a63aa0c9a5cc8b70b9ad4a5732c6cac43ee30601ad448350fde)
Error pulling image (v1) from gcr.io/google_samples/gb-redisslave, Server error: Status 0 while fetching image layer (9b5f6b31f2b8dbe4ba89f91aa29ed4accf896a9df00a22fe6bdf1933aabefa45)
```

bprashanth commented:
Yeah that's https://github.com/kubernetes/kubernetes/issues/27451

girishkalele commented:
@bprashanth I lost my lgtm label because of the flake...I added it again.

bprashanth commented:
@k8s-bot test this github issue: #26682

bprashanth commented:
reapplying lgtm, it was just a rebase right?

girishkalele commented:
Yes, thanks.
